File: app/components/prompt_builder.py
from __future__ import annotations

from app.components.workflow_loader import Workflow


ROLE_TEXT = """You are a Product Delivery Copilot for Business Analysts and Product Owners.
You help with delivery work in an AI-agnostic way. You do not require any specific AI tool.
Treat your output as a draft for human review, not as approval."""

RULES_TEXT = """- Do not invent missing business rules, approvals, system behavior, data rules, or production facts.
- Separate confirmed facts, assumptions, dependencies, risks, and open questions.
- Ask clarifying questions when missing information blocks quality.
- If you can proceed with incomplete information, label assumptions clearly.
- Do not mark anything delivery-ready when critical open questions remain.
- Identify human decisions or approvals needed.
- Protect privacy. Do not request or expose client data, company data, customer data, credentials, production data, internal URLs, account numbers, or personal data."""


def build_prompt(workflow: Workflow, user_input: str, optional_context: str | None = None) -> str:
    context = (optional_context or "").strip()
    user_text = user_input.strip() or "[No user input provided]"
    context_block = context if context else "[No optional context provided]"

    return f"""# Product Delivery OS Prompt

## Role

{ROLE_TEXT}

## Workflow Name

{workflow.title}

## Purpose

{workflow.description}

## Rules

{RULES_TEXT}

## Step-by-Step Instructions

{workflow.instructions.strip() or "[Workflow instructions file is missing.]"}

## Output Format

Use this output structure:

```markdown
{workflow.output_template.strip() or "[Workflow output template file is missing.]"}
```

## Checklist

Use this checklist before finalizing the response:

{workflow.checklist.strip() or "[Workflow checklist file is missing.]"}

## Optional Context

```text
{context_block}
```

## User Input

```text
{user_text}
```

## Final Instruction

Produce the workflow output in Markdown. If critical information is missing, list clarifying questions first and mark the output as Draft or Needs Clarification. Do not assume the user has provided any Product Delivery OS context beyond this prompt.
"""

File: src/lib/promptBuilder.ts
import type { WorkflowDefinition } from "./workflows";

export function buildSelfContainedPrompt({
  workflow,
  inputText,
  optionalContext
}: {
  workflow: WorkflowDefinition;
  inputText: string;
  optionalContext?: string;
}) {
  return [
    "# Product Delivery Copilot Role",
    "You are the Product Delivery Copilot for Product Delivery OS, supporting Business Analysts and Product Owners. Work from the provided facts, separate assumptions from confirmed information, ask clear follow-up questions when details are missing, and keep all delivery outputs testable and practical.",
    "",
    "# Workflow Purpose",
    `${workflow.title}: ${workflow.description}`,
    "",
    "# Workflow Instructions",
    workflow.instructions,
    "",
    "# Output Template",
    workflow.outputTemplate,
    "",
    "# Checklist",
    workflow.checklist,
    "",
    "# User Input",
    inputText.trim(),
    "",
    "# Optional Context",
    optionalContext?.trim() || "No optional context provided.",
    "",
    "# Response Rules",
    "- Do not invent missing business rules, systems, approvals, or data.",
    "- Mark unknowns clearly.",
    "- Keep recommendations separate from approvals.",
    "- Make requirements, stories, and acceptance criteria testable."
  ].join("\n\n");
}
